core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala

```diff
@@ -398,40 +414,96 @@
     }
 
 
-  final def updateWorkspaceRecord: MigrateAction[Unit] =
+  final def restoreIamPoliciesAndUpdateWorkspaceRecord: MigrateAction[Unit] =
     withMigration(_.tmpBucketDeleted.isDefined) { (migration, workspace) =>
       for {
-        _ <- MigrateAction.raiseWhen(migration.newGoogleProjectId.isEmpty || migration.newGoogleProjectNumber.isEmpty) {
-          noGoogleProjectError(migration, workspace)
+        (workspaceService, sam, userInfo) <- MigrateAction.asks { env =>
+          (env.workspaceService, env.sam, env.userInfo)
         }
-        _ <- inTransaction(dataAccess =>
-          dataAccess
-            .workspaceQuery
+
+        googleProjectId = migration.newGoogleProjectId.getOrElse(
+          throw noGoogleProjectError(migration, workspace)
+        )
+
+        _ <- MigrateAction.fromFuture {
+          for {
+            accessPolicies <- sam
+              .listPoliciesForResource(
+                SamResourceTypeNames.workspace,
+                workspace.workspaceId,
+                userInfo
+              )
+              .map(_.map(p => p.policyName -> p.email).toMap)
+
+            // billing project owners get special billing project owner-y iam roles on the
+            // workspace's google project. See the link below for more info.
+            // https://docs.google.com/document/d/1uOGtmw_l_Ve2gqJohLGntJmQz69md2h3T2EAvyKTm20
+            billingProjectOwnerPolicyEmail = accessPolicies
+              .getOrElse(projectOwner, throw WorkspaceMigrationException(
+                message = s"""Workspace migration failed: no "$projectOwner" policy on workspace.""",
+                data = Map("migrationId" -> migration.id, "workspace" -> workspace.toWorkspaceName)
+              ))
+
+            _ <- workspaceService.setupGoogleProjectIam(
+              googleProjectId,
+              accessPolicies,
+              billingProjectOwnerPolicyEmail
+            )
+          } yield accessPolicies
+        }
+
+        // Now we'll update the workspace record with the new google project id.
+        // Why? Because the WorkspaceService does it in this order when creating the workspace.
+        _ <- inTransaction {
+          _.workspaceQuery
             .filter(_.id === workspace.workspaceIdAsUUID)
             .map(w => (w.googleProjectId, w.googleProjectNumber))
-            .update((migration.newGoogleProjectId.get.toString, migration.newGoogleProjectNumber.map(_.toString)))
-        )
+            .update((googleProjectId.value, migration.newGoogleProjectNumber.map(_.toString)))
+        }
+
+        storageService <- MigrateAction.asks(_.storageService)
+        _ <- MigrateAction.fromFuture {
```

Comment:
Again, my preference would be to extract the right side of this "gets" into its own method.